File: mirror/tier2-dart/showcase/pattern_performance.dart
// MIRROR V3: Tier 2 Dart Showcase
// Category: Pattern Matching
// Module: pattern_performance.dart

String fastMatch(dynamic value) {
  if (value is int) return 'fastInt';
  if (value is String) return 'fastStr';
  if (value is List) return 'fastList';
  if (value is Map) return 'fastMap';
  return 'other';
}

dynamic optimizedDispatch(dynamic value, String operation) {
  if (value is int) {
    switch (operation) {
      case 'double': return value * 2;
      default: return null;
    }
  } else if (value is String) {
    switch (operation) {
      case 'upper': return value.toUpperCase();
      default: return null;
    }
  } else if (value is List) {
    switch (operation) {
      case 'length': return value.length;
      default: return null;
    }
  }
  return null;
}

Map<String, dynamic> batchMatch(List<dynamic> values) {
  final results = <String>[];
  final typeCache = <String, int>{};
  
  for (final v in values) {
    final t = v.runtimeType.toString();
    typeCache[t] = (typeCache[t] ?? 0) + 1;
    results.add(fastMatch(v));
  }
  
  return {'matches': results, 'typeCounts': typeCache};
}

void main() {
  final result = <String, dynamic>{
    'fastInt': fastMatch(42),
    'fastStr': fastMatch('hello'),
    'fastList': fastMatch([1, 2, 3]),
    'dispatchDouble': optimizedDispatch(5, 'double'),
    'dispatchUpper': optimizedDispatch('test', 'upper'),
    'batchCount': batchMatch([1, 'a', 2, 'b', 3])['matches'],
  };
  print(result);
}

File: mirror/tier2-ruby/showcase/pattern_performance.rb
# MIRROR V3: Tier 2 Ruby Showcase
# Category: Pattern Matching
# Module: pattern_performance.rb
# Purpose: Demonstrate optimized pattern matching

def fast_match(value)
  case value
  when Integer
    "fast_int"
  when String
    "fast_str"
  when Array
    "fast_list"
  when Hash
    "fast_hash"
  else
    "other"
  end
end

def optimized_dispatch(value, operation)
  case value
  when Integer
    case operation
    when "double"
      value * 2
    end
  when String
    case operation
    when "upper"
      value.upcase
    end
  when Array
    case operation
    when "length"
      value.size
    end
  end
end

def batch_match(values)
  results = []
  type_cache = {}
  
  values.each do |v|
    t = v.class.name
    type_cache[t] ||= 0
    type_cache[t] += 1
    results << fast_match(v)
  end
  
  {matches: results, type_counts: type_cache}
end

result = {
  fast_int: fast_match(42),
  fast_str: fast_match("hello"),
  fast_list: fast_match([1, 2, 3]),
  dispatch_double: optimized_dispatch(5, "double"),
  dispatch_upper: optimized_dispatch("test", "upper"),
  batch_count: batch_match([1, "a", 2, "b", 3])[:matches].size,
}

puts result.inspect
